File: terminal_app/internal/regen/schedule.go
package regen

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

// ParseClock liest eine Uhrzeit im Format "HH:MM".
func ParseClock(s string) (hour, minute int, err error) {
	parts := strings.Split(strings.TrimSpace(s), ":")
	if len(parts) != 2 {
		return 0, 0, fmt.Errorf("time %q: expected format HH:MM", s)
	}
	hour, err = strconv.Atoi(strings.TrimSpace(parts[0]))
	if err != nil || hour < 0 || hour > 23 {
		return 0, 0, fmt.Errorf("time %q: hour must be between 0 and 23", s)
	}
	minute, err = strconv.Atoi(strings.TrimSpace(parts[1]))
	if err != nil || minute < 0 || minute > 59 {
		return 0, 0, fmt.Errorf("time %q: minute must be between 0 and 59", s)
	}
	return hour, minute, nil
}

// AtClock legt eine Uhrzeit auf denselben Kalendertag wie ref (in dessen Zone).
func AtClock(ref time.Time, hour, minute int) time.Time {
	y, m, d := ref.Date()
	return time.Date(y, m, d, hour, minute, 0, 0, ref.Location())
}

// NextOccurrence liefert das nächste Auftreten von hour:minute nach base.
// Liegt die Uhrzeit heute schon hinter base, wird auf morgen gerollt.
//
// Der Tageswechsel läuft über AddDate, damit über eine Zeitumstellung hinweg
// die Wanduhrzeit erhalten bleibt: 14:05 bleibt 14:05, auch wenn der Tag 23
// oder 25 Stunden hat.
func NextOccurrence(base time.Time, hour, minute int) time.Time {
	t := AtClock(base, hour, minute)
	if !t.After(base) {
		t = AtClock(base.AddDate(0, 0, 1), hour, minute)
	}
	return t
}

// NextWholeHourUTC ist der Fallback-Tick-Anchor, wenn die API nicht erreichbar
// ist: WarEra tickt zur vollen Stunde UTC.
func NextWholeHourUTC(now time.Time) time.Time {
	u := now.UTC()
	return time.Date(u.Year(), u.Month(), u.Day(), u.Hour(), 0, 0, 0, time.UTC).Add(time.Hour)
}

File: web_app/src/lib/schedule.ts
/**
 * Uhrzeiten lesen und auf das nächste Auftreten legen.
 *
 * Übersetzung von `internal/regen/schedule.go`, mit den Zonen-Helfern aus
 * `zone.ts` an der Stelle, an der Go eine `*time.Location` benutzt.
 */
import { addWallDays, atClock } from './zone'

/** Eine Uhrzeit ohne Datum. */
export interface Clock {
  hour: number
  minute: number
}

/**
 * Liest eine Uhrzeit im Format „HH:MM“. `null` heißt: nicht lesbar — die
 * Oberfläche meldet das über `err.clock`, deshalb braucht es hier keinen
 * eigenen Fehlertext.
 */
export function parseClock(s: string): Clock | null {
  const parts = s.trim().split(':')
  if (parts.length !== 2) return null
  const [rawHour, rawMinute] = parts as [string, string]
  if (!/^\d{1,2}$/.test(rawHour.trim()) || !/^\d{1,2}$/.test(rawMinute.trim())) return null
  const hour = Number(rawHour.trim())
  const minute = Number(rawMinute.trim())
  if (hour < 0 || hour > 23 || minute < 0 || minute > 59) return null
  return { hour, minute }
}

/** Formatiert eine Uhrzeit wieder als „HH:MM“. */
export function formatClockValue(c: Clock): string {
  return `${String(c.hour).padStart(2, '0')}:${String(c.minute).padStart(2, '0')}`
}

/**
 * Das nächste Auftreten von `hour:minute` nach `base`. Liegt die Uhrzeit heute
 * schon hinter `base`, wird auf morgen gerollt.
 *
 * Der Tageswechsel läuft kalendarisch, damit über eine Zeitumstellung hinweg
 * die Wanduhrzeit erhalten bleibt: 14:05 bleibt 14:05, auch wenn der Tag 23
 * oder 25 Stunden hat.
 */
export function nextOccurrence(base: number, zone: string, hour: number, minute: number): number {
  const today = atClock(base, zone, hour, minute)
  if (today > base) return today
  return atClock(addWallDays(base, zone, 1), zone, hour, minute)
}

/**
 * Der Fallback-Tick-Anchor, wenn die API nicht erreichbar ist: WarEra tickt zur
 * vollen Stunde UTC.
 */
export function nextWholeHourUTC(now: number): number {
  const d = new Date(now)
  return Date.UTC(d.getUTCFullYear(), d.getUTCMonth(), d.getUTCDate(), d.getUTCHours()) + 60 * 60 * 1000
}
